src/hotspot/share/opto/addnode.hpp

```diff
@@ -172,6 +173,23 @@ class AddPNode : public Node {
                                      // second return value:
                                      intptr_t& offset);
 
+  // Helper methods roughly modeled after GraphKit:
+  static Node* make(PhaseIterGVN& igvn, Node* base, Node* ptr, Node* offset) {
+    return igvn.register_new_node_with_optimizer(new AddPNode(base, ptr, offset));
+  }
+
+  static Node* make(PhaseIterGVN& igvn, Node* base, int offset) {
```

Comment:
The overloads do look confusing now. When seeing a call to a factory method, I expect to see all inputs to be separately initialized. With `basic_plus_adr` the intention was clearer.

I'm not against the refactoring, but it feels like it moves the code in the wrong direction. The root problem is that we can't use `GraphKit `during macro expansion, so we duplicated functionality there over time. (`PhaseIdealLoop` is another victim.)

A better way would be to use `GraphKit` directly. And when we discussed it last time [1], Roland mentioned it was already implemented in Valhalla repo. 

[1] https://mail.openjdk.org/pipermail/hotspot-compiler-dev/2020-August/039612.html

Reply:
```
static Node* make(PhaseIterGVN& igvn, Node* base, Node* ptr, Node* offset) {
    return igvn.register_new_node_with_optimizer(new AddPNode(base, ptr, offset));
  }
```
vs
```
  static Node* make(PhaseIterGVN& igvn, Node* base, int offset) {
    return (offset == 0) ? base : make(igvn, base, base, (Node*) igvn.makecon(offset));
  }
```

IMO latter case adds more confusion than improves readability. Why not  unconditionally call `AddPNode::make(base, base, offset)` instead?

Reply:
Thanks for clarifying @iwanowww . I think I see your point now. My original intent was to just move these methods out of PhaseMacroExpand and not much else. 

I'm going to do some more refactoring and patch all users of these make methods to just use this single method: `static Node* make(PhaseIterGVN& igvn, Node* base, Node* ptr, Node* offset)`. What do you think?

Reply:
Sounds good. 

But in the future I'd like to see `PhaseMacroExpand` and `PhaseIdealLoop` migrated to `GraphKit` instead.

Reply:
Actually, I think it's silly to create a method to just do a "return ... new AddP":
```
static Node* make(PhaseIterGVN& igvn, Node* base, Node* ptr, Node* offset) {
    return igvn.register_new_node_with_optimizer(new AddPNode(base, ptr, offset));
}
```
I can just replace the uses of this `make` with a `new AddP...`.

> But in the future I'd like to see PhaseMacroExpand and PhaseIdealLoop migrated to GraphKit instead.

Roger that.

Reply:
@iwanowww do you think I should just withdraw this PR and close the associated RFE?

Reply:
It's up to you, Cesar. I find existing code good enough, but also I don't have anything against your proposal.

Reply:
Thanks, Vladimir. Since the work is already done and it may improve things slightly, I'll keep the PR open.
I created this RFE for migrating PhaseMacroExpand, PhaseIdealLoop to GraphKit: https://bugs.openjdk.org/browse/JDK-8316560